python/ray/log.py

```diff
@@ -0,0 +1,195 @@
+        """Add context information to the log record.
+
+        This filter adds a package name from where the message was generated as
+        well as the worker IP address, if applicable.
+
+        Args:
+            record: Record to be filtered
+
+        Returns:
+            True if the record is to be logged, False otherwise. (This filter only
+            adds context, so records are always logged.)
+        """
+        match = self.logger_regex.search(record.name)
+        if match and match["subpackage"] in self.package_message_names:
+            record.package = f"[Ray {self.package_message_names[match['subpackage']]}]"
+        else:
+            record.package = ""
+
+        # Lazily load ray when this is called to avoid circular import
+        import ray
+
+        if hasattr(ray, "_private"):
+            if self.global_worker_mode is None:
+                self.global_worker_mode = ray._private.worker.global_worker.mode
+
+            if (
+                self.global_worker_mode
+                and self.global_worker_mode == ray._private.worker.WORKER_MODE
+            ):
+                record.node_ip_address = (
+                    ray.runtime_context.get_runtime_context().worker.node_ip_address
+                )
+
+        return True
+
+
+class AnsiStripFormatter(logging.Formatter):
+    """Formatter which strips ANSI escape codes from log messages.
+
+    These escape codes produce noise in output files and conflict with logging,
+    and so are stripped here.
+    """
+
+    strip_ansi_regex = re.compile(r"\x1b\[[0-9;]*m")
+
+    def format(self, record):
+        return re.sub(self.strip_ansi_regex, "", super().format(record))
+
+
+class PlainRayHandler(logging.StreamHandler):
+    """A plain log handler."""
+
+    def __init__(self):
+        super().__init__(stream=sys.stdout)
+
+        # Needed because the existing logging system causes workers to write to
+        # a file that is read by the driver. This will need to be removed when
+        # a proper logging system is set up in a way that lets workers pass
+        # LogRecords to the driver.
+        self.plain_handler = logging.StreamHandler()
+        self.plain_handler.level = self.level
+        self.plain_handler.formatter = logging.Formatter(fmt="%(message)s")
+
+    def emit(self, record: logging.LogRecord):
+        """Emit the log message.
+
+        If this is a worker, serialize the record and send it to the driver via TCP.
+        If this is the driver, emit the message using the appropriate console handler.
+
+        Args:
+            record: Log record to be emitted
+        """
+        import ray
+
+        if (
+            hasattr(ray, "_private")
+            and ray._private.worker.global_worker.mode
+            == ray._private.worker.WORKER_MODE
+        ):
+            self.plain_handler.emit(record)
+        else:
+            logging.StreamHandler.emit(self, record)
+
+
+def call_once(func: Callable[[None], None]):
```

Comment:
hmm why not just using global variable ?

Reply:
logger_initialized = False

def f():
    global logger_initialized
    if logger_initialized:
        return

Reply:
Well, I was thinking that a decorator would be more reusable, and because of concerns about polluting the namespace with global stuff. It also encapsulated the reconfiguration debug message nicely. I'm happy to change this though if there's already an established precedent.